src/borg/cache.py

```diff
@@ -452,25 +455,34 @@ def chunk_decref(self, id, stats):
         else:
             stats.update(-size, -csize, False)
 
-    def file_known_and_unchanged(self, path_hash, st, ignore_inode=False):
+    def file_cache_key(self, hash, path, st, ignore_inode):
+        if ignore_inode or fstype(path) not in FS_WITH_STABLE_INODES:
```

Comment:
the fstype() call could be done less frequently (only at fs boundaries) and the result kept somewhere.